# Embeddings

This lab is designed to help you solidify your understanding of embeddings by applying them to tasks like semantic similarity, clustering, and building a semantic search system.

### Tasks:
- Task 1: Semantic Similarity Comparison
- Task 2: Document Clustering
- Task 3: Enhance the Semantic Search System


## Task 1: Semantic Similarity Comparison
### Objective:
Compare semantic similarity between pairs of sentences using cosine similarity and embeddings.

### Steps:
1. Load a pre-trained Sentence Transformer model.
2. Encode the sentence pairs.
3. Compute cosine similarity for each pair.

### Dataset:
- "A dog is playing in the park." vs. "A dog is running in a field."
- "I love pizza." vs. "I enjoy ice cream."
- "What is AI?" vs. "How does a computer learn?"


In [5]:
from sentence_transformers import SentenceTransformer
from sklearn.metrics.pairwise import cosine_similarity

# Load pre-trained model
model = SentenceTransformer('all-MiniLM-L6-v2')

# Sentence pairs
sentence_pairs = [
    ("A dog is playing in the park.", "A dog is running in a field."),
    ("I love pizza.", "I enjoy ice cream."),
    ("What is AI?", "How does a computer learn?")
]

# Compute similarities

#YOUR CODE HERE

# Encode sentences
embeddings = [model.encode([pair[0], pair[1]]) for pair in sentence_pairs]

# Compute cosine similarities
similarities = [cosine_similarity([emb[0]], [emb[1]])[0][0] for emb in embeddings]

# Print results
for i, (pair, similarity) in enumerate(zip(sentence_pairs, similarities)):
    print(f"Pair {i+1}: {pair[0]} | {pair[1]}")
    print(f"Cosine Similarity: {similarity:.4f}\n")


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

C:\Users\isabe\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.12_qbz5n2kfra8p0\LocalCache\local-packages\Python312\site-packages\huggingface_hub\file_download.py:140: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\isabe\.cache\huggingface\hub\models--sentence-transformers--all-MiniLM-L6-v2. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.7k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

Pair 1: A dog is playing in the park. | A dog is running in a field.
Cosine Similarity: 0.5220

Pair 2: I love pizza. | I enjoy ice cream.
Cosine Similarity: 0.5281

Pair 3: What is AI? | How does a computer learn?
Cosine Similarity: 0.3194



### Questions:
- Which sentence pairs are the most semantically similar? Why?
- Can you think of cases where cosine similarity might fail to capture true semantic meaning?


1. Which sentence pairs are the most semantically similar? Why?

The most semantically similar sentence pair is "I love pizza." vs. "I enjoy ice cream." with a cosine similarity of 0.5281. This is because both sentences express a positive emotion (love/enjoy) towards food items, sharing a similar structure and intent. Although the specific objects (pizza and ice cream) differ, the semantic relationship is close due to shared sentiment and context.

2. Can you think of cases where cosine similarity might fail to capture true semantic meaning?

Yes, cosine similarity might fail in cases such as : Sentences like "The weather is bad" vs. "It’s raining heavily" may have low cosine similarity despite expressing a similar idea; Sentences such as "Dog bites man" vs. "Man bites dog" would have high similarity due to shared vocabulary, but their meanings are entirely different; In technical or domain-specific contexts, embeddings might fail if the pre-trained model isn’t specialized, e.g., comparing "The model overfits" vs. "The algorithm is robust" in machine learning and sentences like "It’s raining cats and dogs" vs. "It’s pouring heavily" may have low similarity despite meaning the same thing.

## Task 2: Document Clustering
### Objective:
Cluster a set of text documents into similar groups based on their embeddings.

### Steps:
1. Encode the documents using Sentence Transformers.
2. Use KMeans clustering to group the documents.
3. Analyze the clusters for semantic meaning.

In [6]:
from sklearn.cluster import KMeans

# Documents to cluster
documents = [
    "What is the capital of France?",
    "How do I bake a chocolate cake?",
    "What is the distance between Earth and Mars?",
    "How do I change a flat tire on a car?",
    "What is the best way to learn Python?",
    "How do I fix a leaky faucet?"
]

# Encode documents

#YOUR CODE HERE


# Encode documents
doc_embeddings = model.encode(documents)



In [7]:
# Perform KMeans clustering

#YOUR CODE HERE
# Perform KMeans clustering
num_clusters = 3  # Define number of clusters
kmeans = KMeans(n_clusters=num_clusters, random_state=42)
kmeans.fit(doc_embeddings)

KMeans(n_clusters=3, random_state=42)

In [8]:
# Print cluster assignments

#YOUR CODE HERE

# Print cluster assignments
for i, doc in enumerate(documents):
    print(f"Document: {doc} | Cluster: {kmeans.labels_[i]}")

Document: What is the capital of France? | Cluster: 2
Document: How do I bake a chocolate cake? | Cluster: 0
Document: What is the distance between Earth and Mars? | Cluster: 0
Document: How do I change a flat tire on a car? | Cluster: 1
Document: What is the best way to learn Python? | Cluster: 2
Document: How do I fix a leaky faucet? | Cluster: 1


### Questions:
- How many clusters make the most sense? Why?
- Examine the documents in each cluster. Are they semantically meaningful? Can you assign a semantic "theme" to each cluster?
- Try this exercise with a larger dataset of your choice

## Task 3: Semantic Search System
### Objective:
Create a semantic search engine:
A user provides a query and you search the dataset for semantically relevant documents to return. Return the top 5 results.

### Dataset:
- Use the following set of documents:
    - "What is the capital of France?"
    - "How do I bake a chocolate cake?"
    - "What is the distance between Earth and Mars?"
    - "How do I change a flat tire on a car?"
    - "What is the best way to learn Python?"
    - "How do I fix a leaky faucet?"
    - "What are the best travel destinations in Europe?"
    - "How do I set up a local server?"
    - "What is quantum computing?"
    - "How do I build a mobile app?"


In [9]:
import numpy as np

# Documents dataset
documents = [
    "What is the capital of France?",
    "How do I bake a chocolate cake?",
    "What is the distance between Earth and Mars?",
    "How do I change a flat tire on a car?",
    "What is the best way to learn Python?",
    "How do I fix a leaky faucet?",
    "What are the best travel destinations in Europe?",
    "How do I set up a local server?",
    "What is quantum computing?",
    "How do I build a mobile app?"
]

# Compute document embeddings
doc_embeddings = model.encode(documents)
#YOUR CODE HERE

In [10]:
# Create the search function
#This function should encode the user query and return the top N documents that most resemble it
def semantic_search(query, documents, doc_embeddings, top_n=5):
    """Returns the top N most similar documents to the query."""
    query_embedding = model.encode([query])
    similarities = cosine_similarity(query_embedding, doc_embeddings)[0]
    top_indices = np.argsort(similarities)[::-1][:top_n]
    
    print(f"Query: {query}\n")
    for i in top_indices:
        print(f"Score: {similarities[i]:.4f} | Document: {documents[i]}")
    print("\n")

In [11]:
# Test the search function
query = "Explain programming languages."
semantic_search(query, documents, doc_embeddings)  

Query: Explain programming languages.

Score: 0.4352 | Document: What is quantum computing?
Score: 0.3188 | Document: What is the best way to learn Python?
Score: 0.1104 | Document: How do I build a mobile app?
Score: 0.0911 | Document: How do I set up a local server?
Score: 0.0906 | Document: What are the best travel destinations in Europe?




### Questions:
- What are the top-ranked results for the given queries?
- How can you improve the ranking explanation for users?
- Try this approach with a larger dataset

1- The most relevant results for "Explain programming languages" are "What is quantum computing?" (0.4352) and "What is the best way to learn Python?" (0.3188), as both relate to computing but do not directly explain programming languages. Other results, such as "How do I build a mobile app?" (0.1104) and "How do I set up a local server?" (0.0911), are somewhat related to development but not specifically to programming languages. "What are the best travel destinations in Europe?" (0.0906) is completely irrelevant but was ranked due to the model’s limitations in accurately capturing semantic meaning.

2-To improve ranking explanations, it is useful to provide justifications by highlighting shared concepts like "programming" or "computing." Incorporating metadata, such as document categories or keywords, can refine the ranking process. Enhancing embeddings or fine-tuning the model for programming-related queries would help capture nuances more accurately. Additionally, applying filtering or reranking can prioritize documents that explicitly match key terms like "programming languages" over general computing topics.